feat(codecs): Implement chunked GELF decoding

jorgehermo9 opened the pull request:
Closes #20769. This PR is kind of large(900 lines of code, 600 are from generated docs), if your prefer to chat via discord (and in order to be more agile while merging this), I'm in the vector community server, username `@jorgehermo9`.

Implementation is based on [Graylog's documentation](https://go2docs.graylog.org/5-0/getting_in_log_data/gelf.html#GELFviaUDP) and [Graylog's go-gelf library](https://github.com/Graylog2/go-gelf/blob/v1/gelf/reader.go)

In my local environment some tests are failing. Could you please trigger the CI so I can see if it is a problem of my environment and if not, I can proceed to fix them?

jorgehermo9 commented on Cargo.toml:
```diff
@@ -139,6 +139,7 @@ serde_json = { version = "1.0.120", default-features = false, features = ["raw_v
 serde = { version = "1.0.204", default-features = false, features = ["alloc", "derive", "rc"] }
 toml = { version = "0.8.14", default-features = false, features = ["display", "parse"] }
 vrl = { version = "0.16.1", features = ["arbitrary", "cli", "test", "test_framework"] }
+tokio = { version = "1.38.0", default-features = false, features = ["full"] }
```
Needed to use tokio inside the `lib/codecs` crate, in order to implement gelf decoding timeouts with tokio tasks.

jorgehermo9 commented on lib/codecs/Cargo.toml:
```diff
@@ -45,6 +46,7 @@ similar-asserts = "1.5.0"
 vector-core = { path = "../vector-core", default-features = false, features = ["vrl", "test"] }
 uuid = { version = "1", default-features = false, features = ["serde", "v4"] }
 rstest = "0.21.0"
+tracing-test = "0.2.5"
```
Found this and it was useful for asserting that tracings logs were produced

I see that this dependency was used in previous versions, such as in vector `0.28.0`, but couldn't find it in the current version. 
Are you okay with using this? The license is APACHE 2.0 OR MIT (https://github.com/dbrgn/tracing-test)

jorgehermo9 replied:
I think it should be added to the licenses csv anyways

jszwedko replied:
👍 you can regenerate them by installing https://github.com/DataDog/rust-license-tool and running `cargo vdev build licenses`.

jorgehermo9 replied:
<img width="794" alt="image" src="https://github.com/user-attachments/assets/8e9b6b06-6220-408a-9a9f-a2ee44933ec8">
<img width="669" alt="image" src="https://github.com/user-attachments/assets/92dc4bd6-1b15-4b22-b4ad-e62a3eab537c">

jorgehermo9 replied:
<img width="783" alt="image" src="https://github.com/user-attachments/assets/c092a632-adc3-481d-8e26-560cb29b2741">
the command does nothing but touching the `Cargo.lock`
<img width="709" alt="image" src="https://github.com/user-attachments/assets/858ecb1f-fb4a-4ffb-b1f5-7161c0defd0b">

I don't know if I'm doing anything wrong...
The dependency is added in the `lib/codecs/Cargo.toml`, not in the workspace root `Cargo.toml`, is `dd-rust-license-tool` aware of dependencies across workspace crates?

jszwedko replied:
Ah, you'll want to run: `cargo vdev build licenses` (or `dd-rust-license-tool write`).

jorgehermo9 replied:
yes, see the second command of the first screenshot! I ran `cargo vdev build licenses` but didn't change any file/output anything, altough dd-rust-license-tool is installed

jszwedko replied:
Oh I see, yes, you are right. It seems like `dd-rust-license-tool` might not be picking up dependencies of subcrates. 🤔

jszwedko replied:
Oh, actually, I think this is because this is a dev-dependency rather than a dependency that is included in the distributed artifacts.

jorgehermo9 replied:
oh, okay!! everything alright then :)

(I'll answer the other comments asap, sorry I had not time these days)

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+    [const { Bytes::new() }; GELF_MAX_TOTAL_CHUNKS as usize];
+const DEFAULT_TIMEOUT_MILLIS: u64 = 5000;
+// TODO: ask what would be an appropriate default value for this
+const DEFAULT_PENDING_MESSAGES_LIMIT: usize = 1000;
```
I don't know what value is appropriate here. Do you have any recommendation?

This limit was enforced so we have a memory-bounded decoder. 

The maximum UDP packet size is 65536 bytes... so with this limit, I think we have roughly 65MB of memory limit for pending messages storage.

However, the framing is agnostic of the transport protocol, so maybe other protocols does not have that per-message size limit and thus this can be "theoretically unbounded" (for example, reading raw bytes from a file).

Should we enforce too a per-message limit such as https://github.com/vectordotdev/vector/blob/1579627ee3c63347fa68fe8ab3e0105bbe6ff3d9/lib/codecs/src/decoding/framing/character_delimited.rs#L51 ?

jorgehermo9 replied:
Addressed this in 85edb0014787b2b1e340078b2a275603b2f0f7f6. Feel free to resolve this thread if the change is what you expected

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+pub struct ChunkedGelfDecoderOptions {
+    /// The timeout in milliseconds for a message to be fully received. If the timeout is reached, the
+    /// decoder will drop all the received chunks of the uncomplete message and start over.
+    ///  The default value is 5 seconds.
```
```suggestion
    /// The default value is 5 seconds.
```

jorgehermo9 replied:
should we document the default value here in a doc-comment?

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+    pub timeout_millis: u64,
+
+    /// The maximum number of pending uncomplete messages. If this limit is reached, the decoder will start
+    /// dropping chunks of new messages. This limit ensures the memory usage of the decoder's state is bounded.
```
The bound is per total messages, but there is no per-message memory usage limit. We can theoretically have a 100GB single message and won't be limited by this setting.

As stated before, should we include a per-message limit?

jszwedko replied:
I like the idea of having a configurable bound on the number of pending messages.

The chunked encoding is only used for UDP, yes? Shouldn't that provide a defacto bound on size? How can we have a 100 GB message?

jorgehermo9 replied:
> The chunked encoding is only used for UDP, yes? 
Yes, it is intended to use it only for UDP and therefore it would be limited by the UDP packets limit of 65KB.
Nevertheless, as the chunked_gelf is a framing method, nothing blocks user to use that method with other types of sources, for example, with `file` sources. Although, it really does not have sense to use that framing method outside of UDP socket sources, and no one will use that in real environments... So maybe it is ok to leave this as it is

jorgehermo9 replied:
Given the nature of gelf's messages, as they are just json, I don't think it would be fine to just truncate the input, as the json message would be most likely broken and the GELF deserialization would fail in nearly all cases after truncating.

Should we instead discard the whole message (including previously stored cunks) if an individual chunk reaches its defined limit or the accumulated chunks limit is reached? I don't know if its worth to do this, but I'm open to implement  it if you see that it would be worth

jszwedko replied:
Ah, yes, I think we'd want to discard messages that exceed the limit (in the future we can route them to a "dead letter" output). This is consistent with the other framers (see example: https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/lib/codecs/src/decoding/framing/character_delimited.rs#L123-L128).

I think it'd be worth it to remove a DOS risk if it isn't too much effort to add.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+/// and [Graylog's go-gelf library](https://github.com/Graylog2/go-gelf/blob/v1/gelf/reader.go).
+#[derive(Debug, Clone)]
+pub struct ChunkedGelfDecoder {
+    // We have to use this decoder to read all the bytes from the buffer first and don't let tokio
```
I don't know if this comment is well explained (or if it has to be removed), tell me what you think please.
That was a bug I faced while testing with big messages. 

we have to rely on the `decode_eof` method for the decoding and not in the `decode` method, such as the `ByteDecoder` does.

In order to now repeat code with that decoder, I used it as an inner decoder since we want the same behaviour; read all bytes from the current buffer.

jszwedko replied:
I think it makes sense. This framer needs to read the entire "packet". As you mentioned in another comment, it may still be prudent to have a configurable `max_length` here to avoid corrupt input ballooning memory usage.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+        let mut state_lock = self.state.lock().unwrap();
+
+        if state_lock.len() >= self.pending_messages_limit {
+            warn!(
```
About all the warnings in this method (not only this one), should we send some about them? Here we are silently discarding messages and I think it would be very useful to see how many packets are being discarded and why.

Currently, I see that the metrics custom-framework is located on `src/internal_events` and there is no easy integration with the crates under `lib/`. Other decoders does not sent any type of metric when discarding, so maybe it does not have sense here either.

jszwedko replied:
As mentioned above, I think we should actually be returning an error in these cases. The caller of the codecs should already be handling the errors by logging and emitting metrics.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+            warn!(message = "Received malformed chunk headers (message ID, sequence number and total chunks) with less than 10 bytes. Ignoring it.",
+                src = src_display,
+                remaining = chunk.remaining(),
+                internal_log_rate_limit = true
```
Added the `internal_log_rate_limit = true` in every log here, I don't know if its okay

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+                let mut state_lock = state.lock().unwrap();
+                if let Some(_) = state_lock.remove(&message_id) {
+                    // TODO: record metrics here? it would be insteresting to know how many messages are being discarded
+                    let message = format!("Message was not fully received within the timeout window of {}ms. Discarding it.",
```
same as before, it would be very nice to know how many messages are timed out...

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+                    // TODO: record metrics here? it would be insteresting to know how many messages are being discarded
+                    let message = format!("Message was not fully received within the timeout window of {}ms. Discarding it.",
+                        timeout.as_millis());
+                    warn!(
```
Should this be a warn, info or debug?

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+        let total_chunks = chunk.get_u8();
+
+        if total_chunks == 0 || total_chunks > GELF_MAX_TOTAL_CHUNKS {
+            warn!(
```
About all the `warn!` logs in this method... it is okay to use that level? Or maybe we should switch to `info` or `debug`?

jszwedko replied:
I actually think we may want to make these `error`s since they result in dropped data. This would typically be done by returning a `BoxedFramingError` from this function where the caller handles logging an error and emitting the appropriate metrics.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+            // otherwise we will have a memory leak due to messages that never complete
+            let state = Arc::clone(&self.state);
+            let timeout = self.timeout.clone();
+            let timeout_handle = tokio::spawn(async move {
```
Following greylog server implementation:
<img width="1359" alt="image" src="https://github.com/user-attachments/assets/18ac054c-1d4c-40e8-b5c2-8494e37fdaf4">
https://go2docs.graylog.org/current/getting_in_log_data/gelf.html#chunked-gelf-messages

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+
+        if message_state.is_chunk_present(sequence_number) {
+            // TODO: add a PR comment asking for info or warn
+            info!(
```
`info` level because a duplicate packet is not really a decoding error, it can be safely ignored (assuming contents were equal, but we are not checking it)

jorgehermo9 replied:
As we are dropping the data, maybe this should also return an `Err`?

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+        message_state.add_chunk(sequence_number, chunk);
+
+        if let Some(message) = message_state.retrieve_message() {
+            state_lock.remove(&message_id);
```
The lock is held until here, but it is not held across `await` calls or anything like that, it should cause no problem although in code lines, it is "far" from the acquisition statement

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+
+    /// Decode a GELF message that may be chunked or not. The source bytes are expected to be
+    /// datagram-based (or message-based), so it must not contain multiple GELF messages
+    /// delimited by '\0', such as it would be in a stream-based protocol.
```
Maybe the last sentence about '\0' delimited is not necessary, but wanted to express it in some way. Opinions?

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+        // the bytes decoder will always return a Ok(None) in this method, but leaving this
+        // here for consistency. Would be better to add a unreachable/panic here if the inner decoder returns
+        // the Some variant?
+        self.bytes_decoder
```
Although we have the `and_then(...)` statement that looks like its doing something... It really doesn't, as in the bytes decoder https://github.com/vectordotdev/vector/blob/1579627ee3c63347fa68fe8ab3e0105bbe6ff3d9/lib/codecs/src/decoding/framing/bytes.rs#L50

It always returns `None`. Although, I added that statement in order to be consistent with the `decode_eof` function of this trait impl, so do not rely in the `BytesDecoder` details... What do you think? We should remove all code in this `decode` impl and just return `Ok(None)` instead?

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+            .and_then(|frame| self.decode_message(frame).transpose())
+            .transpose()
+    }
+    fn decode_eof(&mut self, buf: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
```
As stated before in the `ChunkedGelfDecoder` struct declaration, we will mainly use the `decode_eof` function for decoding, as we have to read all the bytes from the buffer for the chunked gelf decoding

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+    }
+
+    #[fixture]
+    fn unchunked_message() -> (BytesMut, String) {
```
What is the correct spelling? I think its correct... should I add the word to the CI checklist?

jszwedko replied:
I think you can just add it to the spellcheck ignore list in `.github/actions/spelling/allow.txt`.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,650 @@
+
+        tokio::time::sleep(Duration::from_millis(DEFAULT_TIMEOUT_MILLIS + 1)).await;
+        assert!(decoder.state.lock().unwrap().is_empty());
+        assert!(logs_contain(
```
Is it okay to assert `log_contains` in these tests? Could not find any easy way of testing that behaviour without doing this

jorgehermo9 commented on lib/codecs/src/decoding/mod.rs:
```diff
@@ -345,6 +361,14 @@ impl DeserializerConfig {
     }
 
+    /// Returns an appropriate default framing config for the given deserializer with message based inputs.
+    pub fn default_message_based_framing(&self) -> FramingConfig {
```
Moved from a static method, as now it should use a `ChunkedGelf` decoder if the deserializer is gelf and the transport protocol is message-based

jorgehermo9 commented on lib/codecs/src/encoding/format/native_json.rs:
```diff
@@ -98,7 +98,12 @@ mod tests {
+        // TODO: remove this comment. Add a PR comment stating that converting the histogram_event to a serde_json::Value
+        // Changed the order of the keys to alphabetical order and the string comparison failed due to the Event struct
+        // not serializing the keys in the same order as the serde_json::Value.
+        assert_eq!(
```
This test was broken for me in the master branch, I don't know if I should remove this change, tell me.

The issue was that `serializer.to_json_value(histogram_event)` returned a `serde_json::Value`, which when serialized with `serde_json::to_string(&json)`, its keys were alphabetically sorted.

When serializing the `histogram_event` with `serializer
            .encode(histogram_event.clone(), &mut bytes)`, the fields are not alphabetically sorted, serde does not enforce this in arbitrary struct serialization (it does on `Value` struct serializations).

I think the test should be change as stated in this diff. Please, tell me your opinions about this

jszwedko replied:
Huh, interesting. It does pass in CI but you are correct that it may be relying on some non-deterministic behavior. The change seems fine to me.

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -152,12 +151,12 @@ impl SourceConfig for SocketConfig {
-                )
-                .build()?;
+                let decoding = config.decoding().clone();
+                let framing = config
```
Now the `framing` depends on the `decoding` that's being used, so the chunked gelf framing is used as default when using the gelf deserializer

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -169,15 +168,12 @@ impl SourceConfig for SocketConfig {
-                    log_namespace,
-                )
-                .build()?;
+                let decoding = config.decoding.clone();
```
In order to be consistent with the UDP mode

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -383,6 +381,53 @@ mod test {
         SourceSender,
     };
 
+    fn get_gelf_payload(message: &str) -> String {
```
I don't know if those test utility methods should be defined here. Didn't find anywhere else to put them.

jszwedko replied:
👍 this is inline with existing patterns in the codebase.

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -383,6 +381,53 @@ mod test {
+        total_chunks: u8,
+        payload: &[u8],
+    ) -> Bytes {
+        const GELF_MAGIC: [u8; 2] = [0x1e, 0x0f];
```
Hardcode the GELF_MAGIC, although we have it in lib/codecs/src/decoding/framing/chunked_gelf.rs#L15, I found weird exporting that constant for testing purposes..

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -383,6 +381,53 @@ mod test {
+        let payload = get_gelf_payload(short_message);
+        let payload_chunks = payload.as_bytes().chunks(max_size).collect::<Vec<_>>();
+        let total_chunks = payload_chunks.len();
+        assert!(total_chunks <= 128, "too many gelf chunks");
```
hardcode the `128` like I did with the `GELF_MAGIC`

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -75,12 +75,11 @@ pub struct UdpConfig {
     #[configurable(derived)]
-    #[serde(default = "default_framing_message_based")]
-    pub(super) framing: FramingConfig,
+    pub(super) framing: Option<FramingConfig>,
```
Changed the `UdpConfig.framing` field to allow Optional framing, so we can compute the framing default based on the deserialized  ( using the function `default_framing_message_based from the `DeserializerConfig`)

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1083,6 +1134,42 @@ mod test {
+    // TODO: maybe this should be in an integration test, such as `src/sources/redis/mod.rs` and `scripts/integration/redis`
+    // and so? There currently are no integration tests for the socket source
+    #[tokio::test]
+    async fn udp_decodes_chunked_gelf_messages() {
```
I don't know if this kind of test is okay to be done here. It is more like an integration test as the ones we have in `scripts/integration` and `tests/integration`. The problem is, there is not any sockets integration test right now and didn't know the best way to do this.

It is kind of weird to test the chunked gelf deserializing in this module... What do you think about this?
If its not such a big deal, maybe we can leave it here, I don't now

jszwedko replied:
Yeah, I see what you are getting at. It would be nice to have integration tests, but you are correct that we don't have any for `socket` yet. Another place we could put them is in `tests/integration`. Those are Rust-based "integration tests" that are more black box.

jorgehermo9 replied:
It seems that there are no sink/source related tests in that directory. I think it feels better as of right now to leave these tests with the other socket tests because they are kind of related...

But I agree with you that these kind of tests are better placed in `tests/integration`, along with a few others in this file that are not from this PR. I think this change would be better placed in a refactor/tech debt PR

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1220,11 +1307,35 @@ mod test {
+    }
+
+    #[cfg(unix)]
+    async fn init_unix_with_config(
```
What to replicate the `init_udp_with_config` and `init_udp_inner` methods for testing with custom configuration (in order to set the gelf deserializer in the config)

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1323,11 +1435,17 @@ mod test {
     ////////////// UNIX DATAGRAM TESTS //////////////
     #[cfg(unix)]
     async fn send_lines_unix_datagram(path: PathBuf, lines: &[&str]) {
+        let packets = lines.into_iter().map(|line| Bytes::from(line.to_string()));
```
as done in the udp part, I had to abstract this utility functions

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1399,6 +1517,45 @@ mod test {
+    // and so? There currently are no integration tests for the socket source
+    #[cfg(unix)]
+    #[tokio::test]
+    async fn unix_datagram_chunked_gelf_messages() {
```
same as the `udp_chunked_gelf_messages` test

jorgehermo9 commented on src/sources/socket/mod.rs:
```diff
@@ -1399,6 +1517,45 @@ mod test {
+            let mut rng = SmallRng::seed_from_u64(seed);
+            // TODO: add a PR comment here stating that in MACOS at least, with bigger messages, the test fails
+            // with the `No buffer space available` error
+            let max_size = 20;
```
(in MacOS at least) When testing with the same limits as the `udp_chunked_gelf_messages` of `max_size = 300` and `.repeat(1000)` it showed an OS error with code 55 and message `No buffer space available` (see [this issue](https://stackoverflow.com/questions/21973661/os-x-udp-send-error-55-no-buffer-space-available) for example).

What do you think?

jszwedko replied:
Huh, that is strange. The lower limits seem ok to me for testing.

jorgehermo9 commented on src/sources/socket/udp.rs:
```diff
@@ -128,7 +127,7 @@ impl UdpConfig {
             port_key: default_port_key(),
             receive_buffer_bytes: None,
-            framing: default_framing_message_based(),
+            framing: None,
```
the default is now calculated when initializing the socket, as done with the `TcpConfig` and `UnixConfig`

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,647 @@
+pub struct ChunkedGelfDecoderOptions {
+    /// The timeout in milliseconds for a message to be fully received. If the timeout is reached, the
+    /// decoder will drop all the received chunks of the uncomplete message and start over.
+    /// The default value is 5 seconds.
```
should we document the default value here in a doc-comment?

jszwedko replied:
I don't think it should be necessary. I believe the generated docs will include the default.

jorgehermo9 commented on website/cue/reference/components/sources/base/amqp.cue:
```diff
@@ -309,6 +309,31 @@ base: components: sources: amqp: configuration: {
+																The default value is 1000.
+																"""
+						required: false
+						type: uint: default: 1000
```
seeing that the default value is here... maybe we should remove it from the description field?

jszwedko replied:
Agreed, let's drop it from the description.

maycmlee commented on website/cue/reference/components/sources/base/amqp.cue:
```diff
@@ -309,6 +309,31 @@ base: components: sources: amqp: configuration: {
 				}
 			}
+			chunked_gelf: {
+				description:   "Options for the chunked gelf decoder."
```
```suggestion
				description:   "Options for the chunked GELF decoder."
```

maycmlee commented on website/cue/reference/components/sources/base/amqp.cue:
```diff
@@ -309,6 +309,31 @@ base: components: sources: amqp: configuration: {
+				type: object: options: {
+					pending_messages_limit: {
+						description: """
+																The maximum number of pending uncomplete messages. If this limit is reached, the decoder will start
+																dropping chunks of new messages. This limit ensures the memory usage of the decoder's state is bounded.
```
```suggestion
																The maximum number of pending incomplete messages. If this limit is reached, the decoder starts
																dropping chunks of new messages. This limit ensures the memory usage of the decoder's state is bounded.
```

maycmlee commented on website/cue/reference/components/sources/base/amqp.cue:
```diff
@@ -309,6 +309,31 @@ base: components: sources: amqp: configuration: {
+					}
+					timeout_millis: {
+						description: """
+																The timeout in milliseconds for a message to be fully received. If the timeout is reached, the
+																decoder will drop all the received chunks of the uncomplete message and start over.
```
```suggestion
																The timeout, in milliseconds, for a message to be fully received. If the timeout is reached, the
																decoder drops all the received chunks of the incomplete message and starts over.
```

maycmlee reviewed (commented):
Just some small suggestions for docs.

jorgehermo9 commented:
thanks @maycmlee! fixing it asap 😁

jszwedko commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,648 @@
+            return Ok(None);
+        }
+
+        let mut state_lock = self.state.lock().unwrap();
```
```suggestion
        let mut state_lock = self.state.lock().expect("poisoned lock");
```

Nit: we prefer to use `expect` rather than `unwrap`.

jszwedko commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,648 @@
+}
+
+#[derive(Debug)]
+struct MessageState {
```
I stumbled across https://docs.rs/timedmap/latest/timedmap/. It seems like that might be a fit here that would let us delete some code?

jszwedko replied:
Granted that seems to be a niche crate so it may not be worth the cost of adding another dependency.

jorgehermo9 replied:
Agree with you!

Furthermore, I think that crate wouldn't allow to have a custom behaviour when evicting the key (i.e the `warn!` log)

jszwedko reviewed (commented):
This is quite a substantial contribution, thanks @jorgehermo9 !

I left a few comments in-line to start, but will need to spend more time with this.

One high-level question I had was: do we need a separate `gelf` and `chunked_gelf` codec? It seems to me that we could just have one `gelf` codec that handles both, no? I see that a message using chunked encoding is indicated by the sender by prepending special bytes (`0x1e 0x0f`) to chunked messages. Otherwise it is a "normal" GELF message.

jorgehermo9 commented:
> I left a few comments in-line to start, but I will need to spend more time with this.

Okay!

> One high-level question I had was: do we need a separate gelf and chunked_gelf codec? It seems to me that we could just have one gelf codec that handles both, no? I see that a message using chunked encoding is indicated by the sender by prepending special bytes (0x1e 0x0f) to chunked messages. Otherwise it is a "normal" GELF message.

Actually, the `chunked_gelf` is not a codec (in the sense of a Vector's codec which is configured with `decoding.codec` and so), although it implements the trait `tokio_util::codec`. It is a framing method such as the `newline_delimited` method, `octec_counting` method and so. Chunked gelf have more aspects of a transport protocol (such as TCP packet buffering and ordering), so I think using it as a `Framer` fits best. This framer takes chunks of `Bytes` and buffers them until it can get a single `Bytes` struct that contains all the bytes for the GELF message. Note that this it is independent of the GELF messages format(`lib/codecs/src/decoding/format/gelf.rs`), what this framer does is to just read the chunked GELF headers and concatenate the chunk's payload into a single "big" payload (this behaviour is very similar of what the other decoders in `lib/codecs/src/decoding/framing` do). This framer supports both chunked gelf messages and self-contained messages (this `if` statement is the one responsible of that https://github.com/vectordotdev/vector/blob/49540d50f4fcaed6801e1cfd2ec4676b14c4d714/lib/codecs/src/decoding/framing/chunked_gelf.rs#L278). Then, that concatenated `Bytes` that forms a single (big) GELF message is passed to the `GelfDeserializer` in order to convert those `Bytes` into a `Event`

Then, the `gelf` codec (`lib/codecs/src/decoding/format/gelf.rs`) `GelfDeserializer` is responsible to parse GELF message `Bytes` as an `Event`, implementing the `Deserializer` trait. This deserializer is agnostic of the protocol used, whether it got the `Bytes` from chunked UDP, from a TCP stream, from a file bytes, etc.

I think is better to separate the framing of the chunked byte (https://github.com/vectordotdev/vector/blob/49540d50f4fcaed6801e1cfd2ec4676b14c4d714/lib/codecs/src/decoding/framing/chunked_gelf.rs#L146) from the GELF message bytes parsing (https://github.com/vectordotdev/vector/blob/49540d50f4fcaed6801e1cfd2ec4676b14c4d714/lib/codecs/src/decoding/format/gelf.rs#L107) and do not include in the GELF deserialize things about chunking and etc, that is more of a transport protocol thing.

I don't know if I'm explaining this propertly, it makes sense in my head. Tell me if you have any more doubts about this

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,715 @@
+}
+
+#[derive(Debug, Snafu)]
+pub enum ChunkedGelfDecoderError {
```
I added these variants for the error handling, instead of logging with `warn!`. Couldn't avoid the logging in the case of the timeouts, since they are done asynchronously.
Maybe the naming could be improved, any feedback?

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,715 @@
+                tokio::time::sleep(timeout).await;
+                let mut state_lock = state.lock().unwrap();
+                if state_lock.remove(&message_id).is_some() {
+                    let message = format!("Message was not fully received within the timeout window of {}ms. Discarding it.",
```
As we are not returning any error from here, there are not recorded metrics about dropped that in this case... is this the correct way to handle this data discard?

jorgehermo9 commented on lib/codecs/src/decoding/framing/mod.rs:
```diff
@@ -34,23 +36,33 @@ use super::StreamDecodingError;
+    /// Coerces the error to a `dyn Any`.
+    /// This is useful for downcasting the error to a concrete type, as we are dealing
+    /// with Box<dyn FramingError> instead of a concrete `FramingError` enum.
+    fn as_any(&self) -> &dyn Any;
```
I had to do this "hack" in order to downcast a `Box<dyn FramingError>` into a concrete `ChunkedGelfDecoderError` in the `chunked_gelf` unit tests.. This is a bit ugly and I don't know if it have sense to change the `BoxedFramingError` into a concrete `FramingError` enum which would have three variants; `std::io::Error`, `LinesCodecError` and `ChunkedGelfDecoderError`... What do you think? Is there a `Box<dyn FramingError>` for a specific reason? using an enum instead would make the tests I added in `decoding::framing::chunked_gelf::tests`simpler and we would avoid vtables

It shouldn't be a big change to do that. I highly prefer to use concrete enum types instead of `Box<dyn Error>`, at least on lib crates

jszwedko replied:
I feel like @bruceg might have thoughts here. Is there a way we could just do implement this in the `chunked_gelf` tests? It seems like this is the first framing method where the tests needed to downcast.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,715 @@
+        )
+    }
+
+    fn downcast_framing_error(error: &Box<dyn FramingError>) -> &ChunkedGelfDecoderError {
```
Had to do this hack in order to match a `Box<dyn FramingError` to a specific `ChunkedGelfDecoderError` enum variant

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,715 @@
+        let frame = decoder.decode_eof(&mut three_chunks[0]);
+        let error = frame.expect_err("Expected an error");
+        let downcasted_error = downcast_framing_error(&error);
+        matches!(
```
I could have done just `matches!(frame, Err(_))`, but then I wouldn't be testing that it returns the concrete enum variant, I think it's better with this "hack"

jorgehermo9 commented:
Addressed the changes you suggested and added some comments explaining them, I think I didn't miss any comment, but tell me if otherwise!

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,715 @@
+        default = "default_timeout_millis",
+        skip_serializing_if = "vector_core::serde::is_default"
+    )]
+    pub timeout_millis: u64,
```
Do you prefer to use `timeout_millis: u64` or `timeout_secs: f64`? How is usually done in this kind of config?

jszwedko replied:
We generally prefer `timeout_secs` with `f64`.

jorgehermo9 commented on lib/codecs/src/decoding/mod.rs:
```diff
@@ -361,6 +378,14 @@ impl DeserializerConfig {
+    /// Returns an appropriate default framing config for the given deserializer with message based inputs.
+    pub fn default_message_based_framing(&self) -> FramingConfig {
+        match self {
+            DeserializerConfig::Gelf(_) => FramingConfig::ChunkedGelf(Default::default()),
```
Where should I document that when setting `decoding.codec=gelf`, `framing.method` defaults to `chunked_gelf`?

jszwedko replied:
Unfortunately I don't think there is a good place to document this at the moment 😞 I think the codec system is complicated enough it'd make sense to have a dedicated page in the docs for it, but that's out-of-scope of this change.

jszwedko commented:
Thanks @jorgehermo9 ! I still have to go through the updates, but thanks for clarifying the distinction between framing and decoding, I see what you are saying: the chunked bit is just about the framing, where the `gelf` codec is still the decoder.

jorgehermo9 commented:
Yes, exactly that!

Don't worry and take your time, I know its a large PR.

jorgehermo9 commented:
Fixed some tests, the CI should pass now

jorgehermo9 commented:
Fixed the clippy lint, didn't ran that in local...

jorgehermo9 commented:
Just a little reminder in case this PR got forgotten @jszwedko. No hurries, but I'd like to know if the direction seems right

jszwedko commented:
> Just a little reminder in case this PR got forgotten @jszwedko. No hurries, but I'd like to know if the direction seems right

Thanks for the bump :) I should be able to make time for this one this week.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,695 @@
+    /// decoder drops all the received chunks of the incomplete message and starts over.
+    #[serde(
+        default = "default_timeout_secs",
+        skip_serializing_if = "vector_core::serde::is_default"
```
I don't know if this `skip_serializing_if` will work as intended

Here, the default is `default_timeout_secs`, which is 5s.
But `f64::default()` is 0s. 

As it is, this will be skipped only if `timeout_secs == 0`, right? 
I saw this pattern in https://github.com/vectordotdev/vector/blob/06ad674ff4fbf2df82c50b787c45316c5b5c8d50/lib/codecs/src/decoding/format/json.rs#L77

jorgehermo9 replied:
Decided to remove the whole `skip_serializing_if` in b214fdb3e23352c5e353529280fda968909b1b5d. If you prefer another approach (for example, to test to the actual default value and skip then), tell me

jorgehermo9 commented:
Ci failed because component docs were not updated. Fixed that, sorry.

jszwedko commented:
Apologies that I still haven't gotten this meaty one. It's on my list!

jorgehermo9 commented:
Dont apologize! I completely understand it. No hurries, really.

Thanks!

jszwedko reviewed (commented):
Apologies for the very long delay in review here! I appreciate you contributing this useful feature.

I'll give the code a read too, to leave some feedback inline, but I had trouble getting this working with fluent-bit which is able to write chunked GELF. What did you use for testing? I seem to be running into an error in deserializing:

```
2024-10-10T21:21:45.017503Z ERROR source{component_kind="source" component_id=source0 component_type=socket}: vector::internal_events::codecs: Failed deserializing frame. error=expected value at line 1 column 1 error_code="decoder_deserialize" error_type="parser_failed" stage="processing" internal_log_rate_limit=true
```

This is what I tried:

fluent.conf
```
[INPUT]
    Name                    tail
    Tag                     log.*
    Path                    /tmp/logs/*.log
    DB                      /tmp/fluent/flb_kube.db
    Mem_Buf_Limit           5MB
    Refresh_Interval        10
    Parser                  json

[OUTPUT]
    Name                    gelf
    Match                   log.*
    Host                    127.0.0.1
    Port                    15514
    Mode                    udp
    Gelf_Short_Message_Key  message
    Compress                false
    Packet_Size             5

[OUTPUT]
    Name                    stdout
    Match                   log.*
```

Note that I configured the packet size to 5.

parsers.conf

```
[PARSER]
    Name        json
    Format      json
    Time_Key    timestamp
    Time_Format %Y-%m-%dT%H:%M:%S %z
```

Running fluent-bit:

```
fluent-bit -c /tmp/fluent.conf -R /tmp/parsers.conf
```

I then used this Vector config:

```yaml
sources:
  source0:
    address: 0.0.0.0:15514
    mode: udp
    type: socket
    decoding:
      codec: gelf
    framing:
      method: chunked_gelf
      chunked_gelf: {} # note that this is required, that might also be a bug
sinks:
  sink0:
    type: console
    inputs:
    - source0
    encoding:
      codec: json
```

Then writing test data:

```
echo '{"host": "localhost", "level": "debug", "message": "1", "time": "2006-07-28T13:22:04Z"}' >> 14.log
```

When using a larger `packet_size`, where the messages aren't chunked, is parsed correctly, but it doesn't seem to be handling the case where the messages are chunked. I'm curious if you can reproduce.

jorgehermo9 commented:
Hi @jszwedko

For testing, I used a custom script that generates the packets, as I didn't found any other way to do it. Thanks for pointing out that fluent-bit config, I will debug what can be happening with that.

The script I used was something like this, please ignore the compression things, as I used it to test the compressed gelf chunks (which are not implemented in this PR and I will submit later)

```python
import os
import socket
import random
import json
import zlib
import gzip

PORT = os.environ.get('PORT', 12201)
HOST = os.environ.get('HOST', '127.0.0.1')
ADDRESS = (HOST, PORT)
MAX_SIZE = 3000
SOCKET = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

def generate_gelf_chunk(message_id, message, seq_num, total_seq):
    return b''.join([
        b'\x1e\x0f',
        message_id,
        seq_num.to_bytes(1, 'big'),
        total_seq.to_bytes(1, 'big'),
        message
    ])

def generate_gelf_chunks(gelf_payload):
    message_id = os.urandom(8)
    chunks = [gelf_payload[i:i+MAX_SIZE] for i in range(0, len(gelf_payload), MAX_SIZE)]
    total_seq = len(chunks)
    chunks = [generate_gelf_chunk(message_id, chunk, i, total_seq) for i, chunk in enumerate(chunks)]
    # Randomly shuffle the chunks
    random.shuffle(chunks)
    return chunks

def generate_gelf_payload(message):
    return json.dumps({
      "version": "1.1",
      "host": "example.org",
      "short_message": message,
      "full_message": "There is no full message bro",
      "timestamp": 1385053862.3072,
      "level": 1,
      "_user_id": 9001,
      "_some_info": "foo",
      "_some_env_var": "bar"
    })


def generate_compressed_gelf_payload(message,compression):
    payload = generate_gelf_payload(message)
    if compression == "zlib":
        return zlib.compress(payload.encode())
    elif compression == "gzip":
        return gzip.compress(payload.encode())
    elif compression == None:
        return payload.encode()
    else:
        raise ValueError("Invalid compression type")


def send_message(message,compression):
    gelf_payload = generate_compressed_gelf_payload(message,compression)
    chunks = generate_gelf_chunks(gelf_payload)
    for chunk in chunks:
        SOCKET.sendto(chunk, ADDRESS)

    print(f"Sent {len(chunks)} chunks")

def main():
    small_message = "This is a small message" *100
    large_message = "This is a large message" * 500
    very_large_message = "This is a very large message" * 10000

    send_message(small_message,None)
    send_message(small_message,None)
    send_message(large_message,None)
    # send_message(very_large_message,none)
    #
    # zlib_message = "This is a zlib message" * 100
    # gzip_message = "This is a gzip message" * 100

    # send_message(zlib_message,"zlib")
    # send_message(gzip_message,"gzip")

if __name__ == '__main__':
    main()
```

jszwedko commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,691 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Nice tests!

jszwedko reviewed (commented):
Alright, left some comments inline too 😄  Apologies again for the long delay in review on this one.

jorgehermo9 commented:
>       chunked_gelf: {} # note that this is required, that might also be a bug

Fixed in https://github.com/vectordotdev/vector/pull/20859/commits/72414d54820436d818ecac49830bb14a47fa1ba6

The failing CI fixed in https://github.com/vectordotdev/vector/pull/20859/commits/6c66ff04aea9e31ce7105f1a28ef45e9f58411e1

----

I tried the fluentbit config that you depicted, but it seems that fluentbit is not encoding the strings as I would expect.

I tested this source with a Graylog Server output stream and it worked too. In the script I used, I think I was complying with the GELF spec, and also in the unit tests added in this PR. The fluentbit behaviour is very weird and I have to spend more time debugging it :thinking: 

Thank you very much @jszwedko for the review and do not worry for the delay!

I will address the comments as soon as I can :smiley:

jorgehermo9 commented:
Debugging a bit about this... It seems that the first bytes sent from fluentbit after the gelf header 
![image](https://github.com/user-attachments/assets/5ac5b38c-c522-476f-a40a-fa51c8b879ac)
Are `0x1f 0x8b 0x08`... Which casually are the gzip headers bytes... (see for example https://stackoverflow.com/a/58552729), so it seems that fluentbit is sending the compressed bytes

I used the same fluentbit config as yours... I think that fluentbit is not using the `Compression false` option...

If I use the code from my compressed gelf branch (still not ready for submitting a PR) https://github.com/jorgehermo9/vector/pull/2

It gets parsed successfully!
![image](https://github.com/user-attachments/assets/96c5c8db-8d56-4c8d-971c-33763274bdc3)

So fluentbit is definitively compressing the payload as gzip 
![image](https://github.com/user-attachments/assets/06ffa830-8c7a-4b8e-ba9f-e16ca9305df5)
https://docs.fluentbit.io/manual/pipeline/outputs/gelf

jszwedko commented:
Aha! You are right, it is compressing even though `Compress` is set to `false`:

https://github.com/fluent/fluent-bit/blob/d94857f43ec2ff2741a0bdf7acd65a796c9132b5/plugins/out_gelf/gelf.c#L209-L226

It seems to compress regardless if the message size is greater than the packet size. I'll open an issue on fluent-bit since I don't see another one.

Good validation for your other PR adding compression support 😄

jszwedko commented:
fluent-bit issue report: https://github.com/fluent/fluent-bit/issues/9482

jorgehermo9 commented:
I absolutely did not expect that my implementation would raise an issue at fluentbit 😄 

I think that the _few_ hours reading a barely documented spec and navigating through various _kind of cursed_ implementations really paid off 😆 

Thanks for submitting the issue!

It seems that as they have the `message_size > packet_size` condition, when there are chunked packets, they will be always compressed. The only situation where they are not compressed is with unchunked messages

nice catch!

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,692 @@
+        &mut self,
+        mut chunk: Bytes,
+    ) -> Result<Option<Bytes>, ChunkedGelfDecoderError> {
+        // Encoding scheme:
```
I think it would be useful to record a metric for tracking the number of current pending messages. How can we approach this? I think this is usually done at Vector's binary level and not in inner libs... as I see a lot of metrics located at `src/internal_events`.

Also, it would be useful to record the number of timed out messages

jszwedko replied:
That would be useful telemetry. Unfortunately I don't think there is a precedent for codecs emitting metrics, so this may require a bit of scaffolding if you want to add it. I don't think we want to emit telemetry directly from the codec crate so it'd look something like exposing metrics that then the caller (Vector) can emit.

I think this could mean adding a trait method to `Framer`, https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/lib/codecs/src/decoding/framing/mod.rs#L67-L81, that exposes telemetry that the caller than polls and emits; or maybe it takes a callback that is called when the codec wants to emit telemetry (example from the [kafka source](https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/src/sources/kafka.rs#L1365-L1369))

jszwedko replied:
And, yes, we generally try to have only `internal_events` actually emit metrics to centralize it all in one place.

jorgehermo9 replied:
I think emitting metrics from this point would require a refactor that in my opinion is slightly off from this PR. Maybe we can create an issue to track metrics in decoders in general.

I don't mind to address it myself as I found it interesting, but I prefer to narrow the scope of this PR

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,735 @@
+    }
+
+    #[tokio::test]
+    async fn decode_shuffled_messages() {
```
included this new test in https://github.com/vectordotdev/vector/pull/20859/commits/2ab9bf3e9d8222ccc6b6b9c7000343469dabc859

jszwedko commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,735 @@
+    // This limitation is due to the fact that the GELF format does not specify the length of the
+    // message, so we have to read all the bytes from the message (datagram)
+    bytes_decoder: BytesDecoder,
+    state: Arc<Mutex<HashMap<u64, MessageState>>>,
```
This would be a performance optimization, but we could think about using `dashmap` here to reduce lock contention. As it stands, it seems like only one packet can be processed at a time due to the locking, which may limit throughput.

jorgehermo9 replied:
I think the udp socket source does not parallelize and the whole process of reading from the socket -> framing -> decoding is entirely sequential from the udp socket's perspective.

I may be wrong but I think https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/src/sources/socket/udp.rs#L177 is purely sequential. 

So I think the lock would only compete with the message timeout acquires, but in that case, if there are a lot of concurrent timeouts, you are right that dashmap couls help

Anyway, I think it would be useful for this last case, but I'd like if you can confirm the first part about the socket's sequential reads

jorgehermo9 replied:
I implemented it with dashmap in https://github.com/jorgehermo9/vector/commit/3e5f57f027ca21709b44758db0b235d527a558a8 at another branch (https://github.com/jorgehermo9/vector/tree/feature/chunked-gelf-dashmap) and it seems that my implementation with it causes deadlocks and several tests fails
![image](https://github.com/user-attachments/assets/b63e7725-b543-4d3f-af93-5a015ab9f4d7)

I'm concerned about the usage of the [entry](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.entry) and [remove](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.remove) methods, which seems to be the cause of the deadlocks. The `message_state` reference that is returned by the `entry` method lives until the end of the scope, but we need to call `remove` in some branches before returning in the `decode_chunk` method, so I think thats the cause of the deadlock. 

I don't know how to model what I'm doing with `DashMap`

jszwedko replied:
Mmm, yeah, you are right that for UDP packets are read and processed sequentially at the moment though would like to improve that in the future (https://github.com/vectordotdev/vector/issues/8518) so this may turn into a bottleneck at that point.

I see for the TCP sockets it uses one decoder instance per stream, same for Unix stream sockets, and for Unix datagram sockets it is similar to UDP where all packets are processed sequentially.

We can go without this for the initial implementation and I can play around with trying to swap in a dashmap after it is merged.

jorgehermo9 replied:
In UDP there is no concept of multiple streams as in TCP, so I think having multiple decoder instances is not worth.

The alternative in this case is to use a worker per message, but I don't think it would be worth as there would be *a lot* of workers doing relatively small work. Maybe we could have a pool of workers (for example, 10, configurable) and use [crossbeam channel](https://docs.rs/crossbeam/latest/crossbeam/channel/index.html) to decouple the socket reads from the decoder's work.

I may give a try to it if we found performance issues the UDP socket (If we add decompression for the GELF chunks, I think this could really be a bottleneck)

jszwedko replied:
> In UDP there is no concept of multiple streams as in TCP

That's true, but we did discuss in https://github.com/vectordotdev/vector/issues/8518 the idea of creating a separate "stream" by grouping together packets from the same source IP / port.

Definitely no need to do this now, but if you are ever interested, https://github.com/DataDog/lading is the tool we use to generate load for benchmarks. We could add the option to generate GELF data there in order to add a a [benchmark experiment to Vector](https://github.com/vectordotdev/vector/tree/master/regression/cases).

jszwedko reviewed (commented):
Thanks @jorgehermo9 ! I left a couple more comments. I think this is looking pretty close.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,855 @@
+    #[serde(default, skip_serializing_if = "vector_core::serde::is_default")]
+    pub pending_messages_limit: Option<usize>,
+
+    /// The maximum length of a single GELF chunk, in bytes. Chunks longer than this length will
```
Please @jszwedko  take a look into the `max_chunk_length` and `max_message_length` documentation. I hope it is well explained

jszwedko replied:
Thanks! I think the descriptions make sense.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,855 @@
+        chunk_length: usize,
+        max_chunk_length: usize,
+    },
+    #[snafu(display("Message with id {message_id} has exceeded the maximum message length and it will be dropped: got {message_length} bytes and max message length is {max_message_length} bytes. Discarding all buffered chunks of that message"))]
```
@jszwedko Note that the discarding notice to the user is done inside this error. It will be logged outside of this framer when the `MaxMessageLengthExceede` error is thrown. This is a bit different from what you depicted about https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/lib/codecs/src/decoding/framing/character_delimited.rs#L123-L128 where the message is silently discarded and no error is returned to the caller.
What do you prefer? To add a discarding warning or an error? In my opinion, this is not really an error and fits more a warning, but as data is discarded, you said in previous comments that we should return an error in those cases.

jszwedko replied:
That's a good question. This has me wondering what the effect of returning an error from `decode()` is. Do you know if it kills the connection? Or if it just emits the error and continues decoding? If it is the latter, than I think returning an error here is appropriate (and we should probably update the other framers to do the same). If it kills the connection, we may want to avoid returning an error unless it is unrecoverable and simply emit warnings 🤔

jorgehermo9 replied:
I don't think the connection is killed. Here https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L175
there is an infinite loop where at each iteration, `socket.recv_from(..)` is called to read an UDP datagram. If any of the `FramedRead` calls returns an error https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L202 
It will be handled by this case
https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L275

But that `FramedRead` stream won't return more than 1 result, as a chunked GELF datagram can't resolve to multiple GELF messages in a single call. The `!error.can_continue()` call is useless in this case, as every datagram would resolve to either: 1 single GELF message (if the chunks that is being processed is the last of the mesasge, no results at all, or an error. In every case, a new iteration of the infinite loop starts again, to read another datagram from the socket at https://github.com/vectordotdev/vector/blob/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d/src/sources/socket/udp.rs#L175C24-L175C40

I hope I'm explaining it well.

TLDR; Does not kill the connection, it continues decoding next datagrams on the socket.

jorgehermo9 commented:
Hi @jszwedko, thank you for your review. Addressed the `max_chunk_length` and `max_message_length` in https://github.com/vectordotdev/vector/pull/20859/commits/2139ae4a9c41b6674dc56851e4a9a955d9345601. I also refactored the tests a little, take a look into that too please.

I have still pending to answer to the metric emitting comment and the `DashMap` thing. I think using `DashMap` won't hurt and it will be easy to do, but I wonder if we would really see a performance improvement.

At first, I think emitting metrics from this point would require a refactor that in my opinion is slightly off from this PR. Maybe we can create an issue to track metrics in decoders in general.

jorgehermo9 commented:
I just notices that I missed to do a `max_message_length` tests. Will do it as soon as I can, but it will be very similar to the `max_chunk_length` one

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,881 @@
+    ///
+    /// This limit takes only into account the chunk's payload and the GELF header bytes are excluded from the calculation.
+    #[serde(skip_serializing_if = "vector_core::serde::is_default")]
+    pub max_chunk_length: Option<usize>,
```
Should we use no limit by default (`None`) Or use the max packet size of UDP datagrams of `65,535` bytes? (actually, we should use `65,535 - 10`, because for this chunk length count we exclude the 10 bytes of the chunked gelf header

jszwedko replied:
Given the framers have a convention of defaulting to no max length, I think we can leave the default for this as `None` too. I also expect people will more often want to limit the total length of the message, rather than each chunk.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,881 @@
+    /// This limit takes only into account the message's payload and the GELF header bytes are excluded from the calculation.
+    /// The message's payload is the concatenation of all the chunks' payloads.
+    #[serde(skip_serializing_if = "vector_core::serde::is_default")]
+    pub max_message_length: Option<usize>,
```
same as above, should we use `None` by default or `65,535 * 128` bytes, which is the max UDP datagram size times the maximum number of chunks.

Both of those defaults would work (theoretically) the same for the UDP socket source.

jszwedko replied:
Let's leave the default as `None` to match the other framers. Could we call this just `max_length`, though, also to match the other framers?

jorgehermo9 replied:
done in https://github.com/vectordotdev/vector/pull/20859/commits/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,881 @@
+
+        if let Some(max_chunk_length) = self.max_chunk_length {
+            let chunk_length = chunk.remaining();
+            ensure!(
```
Should we remove the state of the message id of this chunk? 

I don't know if it makes sense to keep the buffered chunks if we discars one, as the message would never complete

jorgehermo9 replied:
I wonder if we should also do this for the `InvalidTotalChunks` error  for example, or in any other case..?

jszwedko replied:
Mmm, yeah, seemingly the message will never be complete in this scenario since one of the chunks is too big. I'd be ok with removing the message id. After thinking about it a bit more, I'm also not sure if users will need to limit the individual chunk size, rather than just the message size, so I'd also be ok if you wanted to drop the chunk max limit for now to simplify things until we need it and just leave max message length.

jszwedko reviewed (commented):
Nice work on this! I left a few more comments / questions.

jorgehermo9 commented:
Hi @jszwedko thank you very much for the review. I think I addressed all that is left in https://github.com/vectordotdev/vector/pull/20859/commits/8fbfeeb8ab6526d56e34f9a30df225a20c008a0d and answer some questions.

jorgehermo9 commented on lib/codecs/src/decoding/framing/chunked_gelf.rs:
```diff
@@ -0,0 +1,821 @@
+    #[serde(default, skip_serializing_if = "vector_core::serde::is_default")]
+    pub pending_messages_limit: Option<usize>,
+
+    /// The maximum length of a single GELF message, in bytes. Messages longer than this length will
```
Is this too verbose? :/

jszwedko replied:
I think it looks good 👍

jszwedko reviewed (approved):
Thanks @jorgehermo9 ! I think this looks good to me now. Thank you for all of your work on this! It was a big one.

Are there any other changes you want to make before we merge this?

jorgehermo9 commented:
Thank you very much for your reviews @jszwedko! It had been a hard one, I appreacite a lot your work on this.

Nothing left on this PR from my side. I'll addressing the decompression part in another PR 😀

jszwedko commented:
@jorgehermo9 just FYI it looks like the tests fail to compile on Windows here 🤔

jorgehermo9 commented:
Hi!

It is so weird that it did not happen before this PR.
It seems that the `GenerateConfig` use statement is missing [here](https://github.com/vectordotdev/vector/actions/runs/11671947624/job/32499494229#step:6:1660)

![image](https://github.com/user-attachments/assets/ce6986e3-31e4-490e-8469-bdf94ec98e39)

And in the code, I can see that it is only imported in linux
![image](https://github.com/user-attachments/assets/838df910-701e-40c2-aac3-0c0c7a0d24b3)

It is surprising that it worked before, I did not touch any of those files :open_mouth:

jorgehermo9 commented:
I think https://github.com/vectordotdev/vector/pull/20859/commits/6f0ec1f24244f6f1f65f3859fac5b32eae254bfd should fix it 

I also removed the `sources::util::net::TcpSource` import, as I think it wasn't being used. I wonder why clippy didn't lint that...

jorgehermo9 commented:
I think https://github.com/vectordotdev/vector/pull/20859/commits/38a2af0621798864d29e73d35bbe19e17415bf47 fixes it. Could you retrigger that action please?

jszwedko reviewed (commented):
/ci-run-component-features

jszwedko commented:
> /ci-run-component-features

Mmm, I forgot this doesn't work on forks. I'll run it locally.
